python/hidet/ir/Grammar.md

```diff
@@ -0,0 +1,101 @@
+Pair := STRING `:` Value
+Tuple := `(` ( Value | $\epsilon$ ) (`,` Value)\* `)`
+List := `[` ( Value | $\epsilon$ ) (`,` Value)\* `]`
+Dict := `{` ( Pair | $\epsilon$ ) (`,` Pair)\* `}`
+
+AttributeName := `#` IDENT
+Attribute := AttributeName `=` Dict `;`
+
+DataType := `i1` | `i8` | `i16` | `i32` | `i64` | `u8` | `u16` | `u32` | `u64` | `f16` | `f32` | `f64` | `void` | `str`
+PtrType := `~`DataType
+TensorType := DataType`<` NAT (`,` NAT)\* (`,` AttributeName | $\epsilon$ ) `>`
+TensorPtrType := `~`TensorType
+Type := DataType | PtrType | TensorType | TensorPtrType
+
+Module := `{` ( Function | Attribute )\* `}` 
+Function := `def` IDENT `(` ( Arg`:` Type | $\epsilon$ ) (`,` Arg`:` Type)\*` )` `->` Type `{` Stmt* `}`
+Stmt := CoreStmt ( AttributeName `;` | $\epsilon$ )
+CoreStmt := DeclareStmt 
+		| BufferStoreStmt 
+		| AssignStmt 
+		| ReturnStmt 
+		| LetStmt 
+		| ForStmt 
+		| WhileStmt
+		| BreakStmt
+		| ContinueStmt
+		| IfStmt
+		| AssertStmt
+		| BlackBoxStmt
+		| AsmStmt
+
+DeclareStmt := `decl` IDENT `:` Type ( `=` Expr | $\epsilon$ ) `;`
+BufferStoreStmt := IDENT `[`Expr`]` `=` Expr `;`
+AssignStmt := IDENT `=` Expr `;`
+ReturnStmt := `return` Expr `;`
+LetStmt := `let` IDENT `:` Type `=` Expr  ( `in` Stmt | $\epsilon$ ) `;`
+ForStmt := `for` ( IDENT | `(` IDENT (`,` IDENT)\* `)` ) `in` Mapping `{` Stmt* `}`
+WhileStmt := `while` Expr `{` Stmt* `}`
+BreakStmt := `break` `;`
+ContinueStmt := `continue` `;`
+IfStmt := `if` Expr `{` Stmt* `}` ( `else` `{` Stmt* `}` | $\epsilon$ )
+AssertStmt := `assert` Expr `;`
+BlackBoxStmt := `BlackBox` `{` STRING `}`
+AsmStmt := ( `volatile` | $\epsilon$ ) `asm` `{ in {` IDENT (`,` IDENT) `}` `out` `{` IDENT (`,` IDENT) `}` `{` STRING `}` `}`
+
+
+BinaryOp := `+` | `*` | `-` | `/` | `//` | `%` | `**` | `^` | `<<` | `>>` | `||` | `&&` | `==` | `!=` | `<` | `>` | `<=` | `>=`
+UnaryOp := `-` | `!` 
+Expr := `(` Expr BinaryOp Expr `)`
+		| `(` UnaryOp Expr `)`
+		| GetItem
+		| TensorSlice
+		| Call
+		| Let
+		| Cast
+		| Constant
+		| IfThenElse
+		| Dereference
+		| Address
+		| Reference
+		| IDENT
+
+GetItem := Expr `[` Expr `]`
+Slice := ( Expr | $\epsilon$ ) `:` ( Expr | $\epsilon$ )
+TensorSlice := Expr `[` Slice (`,` Slice)\* `]`
+Call := IDENT `(` ( Expr | $\epsilon$ ) (`,` Expr)\* `)`
+Let := `let` `(` IDENT `:` Type `=` Expr `)` `in` `(` Expr `)`
+Cast := IDENT `as` Type
+Constant := BOOl | INT | FLOAT
+IfThenElse := `if` Expr `then` `{` Expr `}` `else` `{` Expr `}`
+Dereference := `deref` Expr
+Address := `addr` Expr
+Reference := `ref` Expr
+
+Mapping := SpatialTaskMapping | RepeatTaskMapping | ComposedTaskMapping
+SpatialTaskMapping := `@spatial` `(` `shape` `=` `[` INT (`,` INT)\* `]` `,` `ranks` `=` `[` INT (`,` INT)\* `]` `)`
+RepeatTaskMapping := `@repeat` `(` `shape` `=` `[` INT (`,` INT)\* `]` `,` `ranks` `=` `[` INT (`,` INT)\* `]` (`,` AttributeName | $\epsilon$ ) `)`
+ComposedTaskMapping := `@compose` `(` Mapping `,` Mapping `)`
+
```

Comment:
Better also add the layout specification.

Reply:
And add it as an optional argument in TensorType. We can opt it if it has row major layout.

Reply:
Layouts can be handled by attributes, which are compile time data.

Reply:
Task mapping is similar like layout. In hidet, we can use expression as the layout/task mapping argument like
```python
def matmul(ptr: ~float32, m: int, n: int):
   a = tensor_pointer(ptr, shape=[m, n], layout=row_major(m, n))
```
One solution is to treat the task mapping and layout construction like normal function calls. In the AST level, we do not distinguish them. In this way, we can have
```
ForMappingStmt := for indices in EXPR `.` `on` `(` EXPR `)`
```

Reply:
Then can we assume that only `SymbolVar` can be parametrized as a layout parameter in a function type signature? 
eg:

```python
import hidet
from hidet.ir.type import tensor_type
from hidet.lang import attrs, f16, tensor_pointer
from hidet.lang.layout import row_major
from hidet.lang.cuda import dynamic_shared_memory

asym = hidet.symbol_var('a', 'int32')
test_layout = row_major(64, asym)

smem_a_type = tensor_type(
        'float16', shape=[64, asym], layout=test_layout
    )

with hidet.script_module() as module:
    @hidet.script
    def test_fn(a: smem_a_type):
        a[0, 0] += 1
    
    @hidet.script
    def main(a: ~f16):
        attrs.cuda.grid_dim = 1
        attrs.cuda.block_dim = 32
        attrs.cuda.dynamic_smem_bytes = 64 * asym
        smem_a = tensor_pointer(
            'float16', shape=[64, asym], layout=test_layout
        )
        smem_a = dynamic_shared_memory(byte_offset=0, dtype=float16)
        test_fn(smem_a)

ir_module = module.ir_module()
print(ir_module)
tst = ir_module.build()
```

Reply:
Better not, it can also be normal Var. If the user want, they can also use expressions.